crates/polars-lazy/src/physical_plan/executors/scan/csv.rs

```diff
@@ -35,21 +38,143 @@ impl CsvExec {
+                }
+
+                let mut n_rows_read = 0usize;
+                let mut out = Vec::with_capacity(self.paths.len());
+                // If we have n_rows or row_index then we need to count how many rows we read, so we need
+                // to delay applying the predicate.
+                let predicate_during_read = predicate
+                    .clone()
+                    .filter(|_| n_rows.is_none() && self.file_options.row_index.is_none());
+
+                for i in 0..self.paths.len() {
+                    let path = &self.paths[i];
+
+                    let df = options_base
+                        .clone()
+                        .with_row_index(self.file_options.row_index.clone().map(|mut ri| {
+                            ri.offset += n_rows_read as IdxSize;
+                            ri
+                        }))
+                        .with_n_rows(n_rows.map(|n| n - n_rows_read))
+                        .try_into_reader_with_file_path(Some(path.clone()))
+                        .unwrap()
+                        ._with_predicate(predicate_during_read.clone())
+                        .finish()?;
+
+                    n_rows_read = n_rows_read.saturating_add(df.height());
+
+                    let df = if predicate.is_some() && predicate_during_read.is_none() {
+                        let predicate = predicate.clone().unwrap();
+
+                        concat_df(
+                            POOL.install(|| {
+                                df.split_chunks()
+                                    .collect::<Vec<_>>()
+                                    .into_par_iter()
+                                    .map(|df| {
+                                        let s = predicate.evaluate_io(&df)?;
+                                        let mask = s
+                                            .bool()
+                                            .expect("filter predicates was not of type boolean");
+                                        df._filter_seq(mask)
+                                    })
+                                    .collect::<PolarsResult<Vec<_>>>()
+                            })?
+                            .iter(),
+                        )?
+                    } else {
+                        df
+                    };
+
+                    out.push(df);
+
+                    if n_rows.is_some() && n_rows_read == n_rows.unwrap() {
+                        if verbose {
+                            eprintln!(
+                                "reached n_rows = {} at file {} / {}",
+                                n_rows.unwrap(),
+                                1 + i,
+                                self.paths.len()
+                            )
+                        }
+                        break;
+                    }
+                }
+
+                concat_df(out.iter())?
```

Comment:
Can we be sure here the the `DataFrame`s have the same `Schema`? 

If so, we can use `concat_df_unchecked`. It isn't unsafe, but it will elide schema checks and duplicate name checks.